### PR TITLE
Account defender support for reCAPTCHA

### PR DESCRIPTION
**Pull request by Xiaoshouzi-gh**
Added account defender support as reCAPTCHA add-on feature.
Added additional integration tests for this feature.

### Discussion

  * Read the contribution guidelines (CONTRIBUTING.md).
  * If this has been discussed in an issue, make sure to link to the issue here. 
    If not, go file an issue about this **before creating a pull request** to discuss.

### Testing

  * Make sure all existing tests in the repository pass after your change.
  * If you fixed a bug or added a feature, add a new test to cover your code.

### API Changes

  * At this time we cannot accept changes that affect the public API.  If you'd like to help 
    us make Firebase APIs better, please propose your change in an issue so that we 
    can discuss it together.

### CODE REVIEW

File: src/utils/error.ts [hunk 1]
```diff
@@ -737,6 +737,10 @@ export class AuthClientErrorCode {
     code: 'invalid-recaptcha-enforcement-state',
     message: 'reCAPTCHA enforcement state must be either "OFF", "AUDIT" or "ENFORCE".'
   }
+  public static RECAPTCHA_NOT_ENABLED = {
+    code: 'racaptcha-not-enabled',
+    message: 'reCAPTCHA enforcement state must be either "AUDIT" or "ENFORCE" to enable account defender.'
+  }
```
Comment (renkelvin) [inline src/utils/error.ts:743]: This error is not specific to account defender. Maybe change the error message to just "reCAPTCHA enterprise is not enabled."

**Review by renkelvin — APPROVED** [review]

File: src/utils/error.ts (hunk 1 above)
Comment (Xiaoshouzi-gh) [reply src/utils/error.ts]: Good call! Fixed it.

File: test/integration/auth.spec.ts [hunk 2]
```diff
@@ -1764,6 +1783,24 @@ describe('admin.auth', () => {
         });
     });
 
+    it('updateTenant() enable Account Defender should be rejected when tenant reCAPTCHA is disabled',
+      function () {
+        // Skipping for now as Emulator resolves this operation, which is not expected.
```
Comment (renkelvin) [inline test/integration/auth.spec.ts:1788]: Maybe add a TODO so someone can fix this in the future.

File: test/integration/auth.spec.ts [hunk 3]
```diff
@@ -1164,29 +1164,41 @@ describe('admin.auth', () => {
...
         emailPasswordEnforcementState:  'OFF',
         managedRules: [{ endScore: 0.1, action: 'BLOCK' }],
+        useAccountDefender: false,
       },
     };
 
     it('updateProjectConfig() should resolve with the updated project config', () => {
       return getAuth().projectConfigManager().updateProjectConfig(projectConfigOption1)
         .then((actualProjectConfig) => {
+          // keys are generated randomly.
```
Comment (renkelvin) [inline test/integration/auth.spec.ts:1200]: Where are the keys generated?

**Review by renkelvin — APPROVED** [review]
LGTM with comments.

File: test/integration/auth.spec.ts (hunk 3 above)
Comment (Xiaoshouzi-gh) [reply test/integration/auth.spec.ts:1200]: this is the response we got from restAPI, and everytime turn on/off the value changes.

File: src/auth/auth-config.ts [hunk 4]
```diff
@@ -1520,17 +1520,25 @@ export interface RecaptchaConfig {
    * The reCAPTCHA keys.
    */
   recaptchaKeys?: RecaptchaKey[];
+
+  /**
+   * Whether to use the account defender for reCAPTCHA assessment.
+   * Initial default value of this is false.
+   */
```
Comment (kevinthecheung) [inline src/auth/auth-config.ts:1527]:
```suggestion
  /**
   * Whether to use account defender for reCAPTCHA assessment.
   * The default value is false.
   */
```

**Review by kevinthecheung — APPROVED** [review]

File: test/unit/auth/project-config.spec.ts [hunk 5]
```diff
@@ -94,6 +96,14 @@ describe('ProjectConfig', () => {
         }).to.throw('"RecaptchaConfig.emailPasswordEnforcementState" must be either "OFF", "AUDIT" or "ENFORCE".');
       });
 
+      it('should throw on non-boolean useAccountDefender attribute', () => {
+        const configOptionsClientRequest = deepCopy(updateProjectConfigRequest) as any;
+        configOptionsClientRequest.recaptchaConfig.useAccountDefender = 'no';
```
Comment (lahirumaramba) [inline test/unit/auth/project-config.spec.ts:101]: Nit: feel free to ignore. You can also test for few more different types in addition to a `string` here.

```
const invalidUseAccountDefender = [null, NaN, 0, 1, '', 'a', [], [1, 'a'], {}, { a: 1 }, _.noop];
    invalidUseAccountDefender.forEach((useAccountDefender) => {
      it(`should throw given invalid useAccountDefender parameter: ${JSON.stringify(useAccountDefender)}`, () => {
```

File: test/unit/auth/tenant.spec.ts [hunk 6]
```diff
@@ -361,6 +371,14 @@ describe('Tenant', () => {
         }).to.throw('"RecaptchaConfig.managedRules" must be an array of valid "RecaptchaManagedRule".');
       });
 
+      it('should throw on non-boolean useAccountDefender attribute', () => {
+        const tenantOptionsClientRequest = deepCopy(clientRequestWithRecaptcha) as any;
+        tenantOptionsClientRequest.recaptchaConfig.useAccountDefender = 'yes';
```
Comment (lahirumaramba) [inline test/unit/auth/tenant.spec.ts:376]: nit: Please see the comment in `test/unit/auth/project-config.spec.ts`

**Review by lahirumaramba — APPROVED** [review]
LGTM. Thanks!

File: src/auth/auth-config.ts (hunk 4 above)
Comment (Xiaoshouzi-gh) [reply src/auth/auth-config.ts]: Done

File: test/unit/auth/tenant.spec.ts (hunk 6 above)
Comment (Xiaoshouzi-gh) [reply test/unit/auth/tenant.spec.ts]: Done

**Pull request merged**